HW5/HW5_pandas_vis.py

```diff
@@ -0,0 +1,499 @@
+# -*- coding: utf-8 -*-
+"""Pandas.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/1dQHJIcbaNMWZbtIr3j61LrjZBJxyunKJ
+
+Полезности:
+
+* Метод merge датафрейма
+* Метод apply серий
+* Метод select_dtypes датафрейма
+* Метод filter датафрейма
+* Методы drop и drop_duplicates датафрейма
+* Метод pivot и pivot_table датафрейма
+* и любые другие
+"""
+
+!pip install -U pip
+!pip install -U matplotlib
+
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import matplotlib
+from matplotlib import rc, rcParams, gridspec
+from matplotlib.patches import ConnectionPatch
+import seaborn as sns
+import math
+
+!git clone -b HW5 --single-branch https://github.com/checheanya/Python_BI_2022
+
+!head -10 ./Python_BI_2022/HW5/data/alignment.bed
+
+!head -10 ./Python_BI_2022/HW5/data/rrna_annotation.gff
+
+"""## Задание 1. Работа с реальными данными (20 баллов)
+
+#### Дано:
+1) аннотация рибосомальной РНК некоторого метагеномного датасета в формате **GFF** (rrna_annotation.gff) \\
+2) файл с выравниванием метагеномной сборки на этот же датасет в формате **BED** 6 (alignment.bed) \\
+
+\\
+#### Задачи:
+1.   Датафреймы
+- конвертация формата в датафрейм
+- колонка с атрибутами - данные о типе рРНК одной короткой строкой (16S, 23S, 5S)
+
+2.   Таблица по хромосомам (геномам) \\
+Для каждой хромосомы (на самом деле это не хромосомы, а референсные геномы) показано количество рРНК каждого типа
+
+3.   Barplot, отображающий эти данные
+
+4.   Таблица успешных РНК \\
+Мы хотим узнать сколько рРНК в процессе сборки успешно собралось. В результате сохранятся только записи об рРНК, интервал которой перекрывался с интервалом контига в выравнивании, это означает, что это ген есть в сборке. Выведите таблицу, содержащую исходные записи об **рРНК** полностью вошедших в сборку (не фрагментом), а также запись о **контиге** в который эта РНК попала. Обратите внимание, что в один контиг может попасть несколько рРНК.
+
+#### File to dataframe
+"""
+
+def read_gff(path):
+    gff_dataframe = pd.read_table(path, names = [
+        "chromosome", "source", "type", "start", "end", "score",
+        "strand", "phase", "attributes"], skiprows = [0])  
+    gff_dataframe["attributes"] = gff_dataframe["attributes"].apply(lambda x: x[5:x.find("S")+1])
```

Comment:
Твой код учитывает только случай, где комментарий занимает только первую строку. Однако чаще всего в файлах данного формата могут быть комментарии на много строк, их нужно учитывать. Благо в пандасе это делается супер просто
```suggestion
    gff_dataframe = pd.read_table(path, names = [
        "chromosome", "source", "type", "start", "end", "score",
        "strand", "phase", "attributes"], comment="#")  
    gff_dataframe["attributes"] = gff_dataframe["attributes"].apply(lambda x: x[5:x.find("S")+1])
```

Reply:
Использование индексов (5 в твоём коде) это крайне ненадёжная и не универсальная штука из-за которой малейшее изменение в данных драмматически сказится на выполнении твоего кода. Если уж использовать какие-то индексы, то вычислять их из данных